src/main/scala/com/criteo/marathon/CapabilitiesPlugin.scala

```diff
@@ -0,0 +1,111 @@
+
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.plugin.plugin.PluginConfiguration
+import mesosphere.marathon.plugin.task._
+import mesosphere.marathon.plugin.{ApplicationSpec, PodSpec}
+import org.apache.mesos.Protos._
+import play.api.libs.json.JsObject
+
+class CapabilitiesPlugin extends RunSpecTaskProcessor with PluginConfiguration with StrictLogging
+{
+  private val MESOS_BOUNDING_CAPABILITIES_LABEL = "MESOS_BOUNDING_CAPABILITIES"
+  private val MESOS_EFFECTIVE_CAPABILITIES_LABEL = "MESOS_EFFECTIVE_CAPABILITIES"
+
+  override def initialize(marathonInfo: Map[String, Any],
+    configuration: JsObject): Unit = {
+  }
+
+  private def capabilityStrToCapability(capabilityStr: String): Option[CapabilityInfo.Capability] = {
+    capabilityStr match {
+      // Only implement network capabilities for now.
+      case "NET_RAW" => Some(CapabilityInfo.Capability.NET_RAW)
+      case "NET_ADMIN" => Some(CapabilityInfo.Capability.NET_ADMIN)
+      case "NET_BIND" => Some(CapabilityInfo.Capability.NET_BIND_SERVICE)
+      case "NET_BROADCAST" => Some(CapabilityInfo.Capability.NET_BROADCAST)
+      case _ => None
+    }
+  }
+
+  private def extractCapabilitiesFromLabels(builder: TaskInfo.Builder, capabilitiesLabelName: String)
+    : (Boolean, Option[CapabilityInfo]) = {
+    import collection.JavaConverters._
+
+    val capabilitiesLabels = builder.getLabels.getLabelsList.stream()
+      .filter(_.getKey.equals(capabilitiesLabelName)).collect(Collectors.toList())
+
+    if (capabilitiesLabels.size() == 0) {
+      return (false, None)
+    }
+
+    if (capabilitiesLabels.size() > 1) {
+      logger.info(s"There should be only exactly one label with key $capabilitiesLabelName for " +
+        s"task ${builder.getTaskId}")
+      return (false, None)
+    }
+
+    val capabilities = capabilitiesLabels.get(0).getValue.split(",")
+      .map(_.trim)
+      .map(capabilityStrToCapability).toSeq
+
+    val unknownCapabilities = capabilities.exists(_.isEmpty)
+
+    val filteredCapabilities = capabilities
+      .filter(_.isDefined)
+      .map(_.get)
+      .asJava
+
+    (unknownCapabilities, Some(CapabilityInfo.newBuilder()
+      .addAllCapabilities(filteredCapabilities).build()))
+  }
+
+  override def taskInfo(appSpec: ApplicationSpec,
+                        builder: TaskInfo.Builder): Unit = {
+    if (!builder.hasLabels) {
+      logger.debug(s"No labels for task ${builder.getTaskId}")
+      return
+    }
+
+    val (unknownBoundingCapabilities, boundingCapabilitiesBuilder) = extractCapabilitiesFromLabels(
+      builder, MESOS_BOUNDING_CAPABILITIES_LABEL)
+    val (unknownEffectiveCapabilities, effectiveCapabilitiesBuilder) = extractCapabilitiesFromLabels(
+      builder, MESOS_EFFECTIVE_CAPABILITIES_LABEL)
+
+    if (unknownBoundingCapabilities || unknownEffectiveCapabilities) {
+      builder.getCommandBuilder.setValue(
+        "echo 'Unknown capabilities. Please double check your configuration.' >&2 && exit 1")
```

Comment:
maybe we should display the capabilities we were unable to recognize?

Reply:
Done